### PR TITLE
Node: URL: correct format/parse parameter types

### PR DESCRIPTION
**Pull request by OliverJAsh**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
- - https://nodejs.org/api/url.html#url_url_format_url_options
- - https://nodejs.org/api/url.html#url_url_parse_urlstring_parsequerystring_slashesdenotehost
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Review by galkin — COMMENTED** [review]
@OliverJAsh, thank you for your work.
I highlight author of affected packages. I will approve your pull after feedback from them.

File: types/node/url.d.ts [hunk 1]
```diff
@@ -1,29 +1,36 @@
...
+        slashes?: boolean | null;
+        port?: string | number | null;
         query?: string | null | ParsedUrlQueryInput;
     }
 
     // Output of `url.parse`
-    interface Url extends UrlObjectCommon {
-        port?: string;
-        query?: string | null | ParsedUrlQuery;
+    interface Url {
```
[the remark this answers is not in the document]
Comment (OliverJAsh) [reply types/node/url.d.ts]: I thought about doing this, but the types are slightly different (e.g. `port` cannot be a `number` here)—so I don't think we could use that here.

File: types/node/url.d.ts [hunk 2]
```diff
@@ -1,29 +1,36 @@
...
+        auth: string | null;
+        hash: string | null;
+        host: string | null;
+        hostname: string | null;
+        href: string | null;
```
Comment (ikokostya) [inline types/node/url.d.ts]: Could you provide case when `href` can be `null`? In the following tests `href` is always `string`:

```sh
$ node --version
v10.16.0
$ node -p 'url.parse("")'
Url {
  protocol: null,
  slashes: null,
  auth: null,
  host: null,
  port: null,
  hostname: null,
  hash: null,
  search: null,
  query: null,
  pathname: null,
  path: null,
  href: '' }
$ nvm use 8
Now using node v8.16.2 (npm v6.4.1)
$ node -p 'url.parse("")'
Url {
  protocol: null,
  slashes: null,
  auth: null,
  host: null,
  port: null,
  hostname: null,
  hash: null,
  search: null,
  query: null,
  pathname: null,
  path: null,
  href: '' }
```

Comment (OliverJAsh) [reply types/node/url.d.ts]: Good catch, I've reverted that particular change.

Comment (galkin) [reply types/node/url.d.ts]: Could you squash your commits into one?

Comment (OliverJAsh) [reply types/node/url.d.ts]: Is that really necessary? I've never had to do that before when contributing to DefinitelyTyped.

Comment (sandersn) [reply types/node/url.d.ts]: Squash merge is the default for DT maintainers so I don't think it's necessary for contributors to squash them ahead of time.

File: types/node/url.d.ts [hunk 3]
```diff
@@ -1,29 +1,36 @@
...
+    interface Url {
+        auth: string | null;
+        hash: string | null;
+        host: string | null;
+        hostname: string | null;
+        href: string;
+        path: string | null;
+        pathname: string | null;
+        protocol: string | null;
+        search: string | null | null;
```
Comment (Huachao) [inline types/node/url.d.ts]: string | null

Comment (ExE-Boss) [reply types/node/url.d.ts]:
```suggestion
	        search: string | null;
```

**Comment by galkin** [comment]
@OliverJAsh, could you make squash for your commits?

**Comment by OliverJAsh** [comment]
@galkin That's not necessary: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40118#discussion_r342116396

**Comment by galkin** [comment]
Thank you i missed the comment above.

**Pull request merged**